### PR TITLE
Reduce duplication of JSON:API code

### PR DESCRIPTION
**Pull request by gabesullice**
There has been a lot of change in JSON:API since a lot of this code would have worked. The changes have been made to make things more encapsulated. So, for example, you don't need to worry about getting includes on only a collection of entities with the same resource type, because all entities are wrapped in a `ResourceObject` class, which carries the resource type for _that_ entity (this means the include resolver won't get confused if you give it a collection of totally random entities). Also, I used a sort of decoration-like pattern in the Cart API controller so that it could copy less code and therefore be less brittle WRT to change.

<hr/>

I'm confused by a couple design decisions, but I'll open separate issues for those things.

### CODE REVIEW

**Comment by gabesullice** [comment]
I should say, the tests don't pass yet. But I think there's some weird, unrelated stuff going on. The tests are ending in a lot of `403 Forbidden` responses when the tests expect 2xx responses. Unfortunately, there's no reasons given by the access results. I'd have to dig in further to figure those out.

File: src/Controller/CartResourceController.php [hunk 1]
```diff
@@ -217,48 +187,27 @@ public function clearItems(OrderInterface $cart) {
...
-        throw new UnprocessableEntityHttpException(sprintf('You must specify a purchasable entity type for row: %s', $key));
-      }
-      if (!isset($order_item_data['purchased_entity_id'])) {
-        throw new UnprocessableEntityHttpException(sprintf('You must specify a purchasable entity ID for row: %s', $key));
-      }
-      if (!$this->entityTypeManager->hasDefinition($order_item_data['purchased_entity_type'])) {
-        throw new UnprocessableEntityHttpException(sprintf('You must specify a valid purchasable entity type for row: %s', $key));
-      }
-    }
+    $resource_type = new ResourceType('virtual_cart', 'virtual_cart', EntityInterface::class);
```
Comment (mglaman) [inline src/Controller/CartResourceController.php:190]: Oh, wait, could we do this and pass `OrderInterface::class` to work around the bundle constraints of JSON API?

**Comment by mglaman** [comment]
I was experiencing a 403 on cart canonical as well. It wasn't working "perfectly", which is one reason I had held off on sending this off to you folks, but this gives me a great step forward to experiment with.

File: src/Controller/CartResourceController.php (hunk 1 above)
Comment (gabesullice) [reply src/Controller/CartResourceController.php:190]: I don't really know where you're going with this

**Comment by mglaman** [comment]
This is probably why 403 occurred. Lacking `view` access.

```
/**
 * Implements hook_entity_access().
 *
 * @todo add this to commerce_cart
 */
function commerce_demo_api_entity_access(EntityInterface $entity, $operation, AccountInterface $account) {
  if ($entity instanceof \Drupal\commerce_order\Entity\OrderInterface) {
    $cart_session = \Drupal::getContainer()->get('commerce_cart.cart_provider');
    if ($operation === 'view') {
      $cart_ids = $cart_session->getCartIds($account);
      $result =  \Drupal\Core\Access\AccessResult::allowedIf(in_array((int) $entity->id(), $cart_ids, TRUE));
      return $result;
    }
  }
}
```

**Comment by mglaman** [comment]
@gabesullice I didn't forget about this. On the way home from MidCamp I worked to get the tests passing _before_ using 2.x-dev (or well, Core, now)

**Pull request merged**